libass/ass_render.c

```diff
@@ -1780,14 +1808,63 @@ wrap_lines_smart(ASS_Renderer *render_priv, double max_text_width)
     }
     text_info->lines[cur_line - 1].len =
         text_info->length - text_info->lines[cur_line - 1].offset;
+}
+
+#ifdef CONFIG_UNIBREAK
+static utf32_t text_info_next_glyph(const void *tinfo, size_t len, size_t *index)
+{
+    const TextInfo *text_info = (const TextInfo*) tinfo;
+    GlyphInfo *cur;
+
+    if (*index >= len)
+        return EOS;
+
+    cur = text_info->glyphs + *index;
+    *index += 1;
+    return cur->symbol;
+}
+#endif
+
+/**
+ * \brief rearrange text between lines
+ * \param max_text_width maximal text line width in pixels
+ * The algo is similar to the one in libvo/sub.c:
+ * 1. Place text, wrapping it when current line is full
+ * 2. Try moving words from the end of a line to the beginning of the next one while it reduces
+ * the difference in lengths between this two lines.
+ * The result may not be optimal, but usually is good enough.
+ *
+ * FIXME: implement style 0 and 3 correctly
+ */
+static void
+wrap_lines_smart(ASS_Renderer *render_priv, double max_text_width)
+{
+    char *unibrks = NULL;
+
+#ifdef CONFIG_UNIBREAK
+    if (render_priv->track->parser_priv->wrap_unicode) {
+        if (render_priv->scratch_buf_len < render_priv->text_info.length) {
+            size_t newlen = FFMAX(2 * render_priv->scratch_buf_len,
+                                  (size_t)render_priv->text_info.length + 20);
```

Comment:
Why `+ 20`?

Style nit: I believe our casts generally have a space after the closing parenthesis.

Reply:
Changed it, thx.